dashboard/app/assets/javascripts/client_state.js

```diff
@@ -42,6 +42,43 @@ dashboard.clientState.reset = function() {
 };
 
 /**
+ * Returns the client-cached copy of the level source for the given
+ * scriptLevelId, if it's newer than the given timestamp.
+ * @param {number} scriptLevelId
+ * @param {number} timestamp
+ */
+dashboard.clientState.sourceForLevel = function (scriptLevelId, timestamp) {
+  var data = localStorage.getItem('source' + scriptLevelId);
+  if (data) {
+    var parsed = JSON.parse(data);
```

Comment:
JSON parsing could fail if we or the browser had a bug or data corruption. How about if we catch the exception and return null in that case.  (This is only supposed to be a soft cache anyway so that's allowed in the contract.)